src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1059 @@
+                                  std::is_same<BsonT, bsoncxx::types::b_bool>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_date>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_int32>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_int64>::value ||
+                                  std::is_same<BsonT, bsoncxx::oid>::value;
+};
+
+class BSONOutputArchive : public OutputArchive<BSONOutputArchive> {
+    /**
+    * The possible states for the BSON nodes being output by the archive.
+    */
+    enum class OutputNodeType { StartObject, InObject, StartArray, InArray };
+
+    typedef bsoncxx::builder::core BSONBuilder;
+
+   public:
+    /**
+    * Construct a BSONOutputArchive that will output to the provided stream.
+    *
+    * @param stream
+    *   The stream to which the archiver will output BSON data.
+    */
+    BSONOutputArchive(std::ostream& stream)
+        : OutputArchive<BSONOutputArchive>{this},
+          _bsonBuilder{false},
+          _writeStream{stream},
+          _nextName{nullptr} {
+    }
+
+   private:
+    /**
+     * Writes the current contents of the BSON document builder to the
+     * output stream.
+     */
+    void writeDoc() {
+        _writeStream.write(reinterpret_cast<const char*>(_bsonBuilder.view_document().data()),
+                           _bsonBuilder.view_document().length());
+        _bsonBuilder.clear();
+    }
+
+   public:
+    /**
+     * Checks if the most recent object written was in the root of the document.
+     * If so, that document is written to the archive.
+     */
+    void writeDocIfRoot() {
+        if (_nodeTypeStack.empty()) {
+            writeDoc();
+        }
+    }
+
+    /**
+     * Starts a new node in the BSON output.
+     *
+     * The node can optionally be given a name by calling setNextName prior to
+     * creating the node. Nodes only need to be started for types that are
+     * objects or arrays.
+     */
+    void startNode() {
+        writeName(true);
+        _nodeTypeStack.push(OutputNodeType::StartObject);
+    }
+
+    /**
+     * Designates the most recently added node as finished.
+     */
+    void finishNode() {
+        // Iff we ended up serializing an empty object or array, writeName
+        // will never have been called - so start and then immediately end
+        // the object/array.
+        //
+        // We'll also end any object/arrays we happen to be in.
+
+        bool closedObj = false;
+        switch (_nodeTypeStack.top()) {
+            case OutputNodeType::StartArray:
+                _bsonBuilder.open_array();
+            case OutputNodeType::InArray:
+                _bsonBuilder.close_array();
+                closedObj = true;
+                break;
+            case OutputNodeType::StartObject:
+                if (_nodeTypeStack.size() > 1) {
```

Comment:
What conditions lead to this case, where the only element on the stack is `StartObject`?